compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp

```diff
@@ -403,23 +403,79 @@ LogicalResult getStaticDims(Operation *op,
   return success();
 }
 
+/// Utility to get the static offsets, sizes and strides for
+/// `AIEX::NpuDmaMemcpyNdOp` with implicit addressing.
+LogicalResult getStaticDimsForImplicitAddressing(
+    Operation *op, MemRefType memrefType, SmallVectorImpl<int64_t> &staticSizes,
+    SmallVectorImpl<int64_t> &staticStrides) {
+  SmallVector<int64_t> shapeArr;
+  for (auto shape : memrefType.getShape()) shapeArr.push_back(shape);
+  int64_t sizeIndex = staticSizes.size() - 1;
+  for (int64_t i = shapeArr.size() - 1; i >= 0; i--) {
+    staticSizes[sizeIndex--] = shapeArr[i];
+  }
+  int64_t strideIndex = staticStrides.size() - 1;
+  // Since aiex.npu.dma_cpy_nd expects 3-dimensional stride attribute, it has
+  // the 4-th implicit dimension as 1. Hence we store that in `prevStride`.
+  int64_t prevStride = 1;
+  for (int64_t i = shapeArr.size() - 1; i >= 0; i--) {
+    staticStrides[strideIndex] = shapeArr[i] * prevStride;
+    prevStride = staticStrides[strideIndex--];
+  }
+  for (int64_t i = strideIndex; i >= 0; i--) {
+    staticStrides[i] = staticStrides[i + 1];
+  }
+  return success();
+}
+
 /// Convert the `amdaie.npu.dma_cpy_nd` operation to `aiex.npu.dma_memcpy_nd`.
 LogicalResult npuDmaCpyNdOpToAIE(IRRewriter &rewriter,
                                  AMDAIE::NpuDmaCpyNdOp dmaOp,
                                  SmallVector<Operation *> &toBeErased,
                                  IRMapping &mapper, IRMapping &bindingsMapper) {
   rewriter.setInsertionPoint(dmaOp);
   // Convert bidirectional `amdaie.npu.dma_cpy_nd` op into two halves.
-  if (dmaOp.hasSourceAddressing()) {
+
+  // TODO(avarma): The logic for checking implicit L3 addressing should just be
+  // part of AMDAIEOps.td. Confirm and address during PR.
+  MemRefType sourceMemrefType =
+      cast<LogicalObjectFifoType>(dmaOp.getDmaCpyNdOp().getSourceType())
+          .getElementType();
+  Attribute sourceMemSpaceAttr = sourceMemrefType.getMemorySpace();
+  MemRefType targetMemrefType =
+      cast<LogicalObjectFifoType>(dmaOp.getDmaCpyNdOp().getTargetType())
+          .getElementType();
+  Attribute targetMemSpaceAttr = targetMemrefType.getMemorySpace();
+  if (dmaOp.hasSourceAddressing() || !sourceMemSpaceAttr) {
+    // DmaOp has source addressing on L3 either explicitly or implicitly.
     SmallVector<Value> empty;
-    SmallVector<int64_t, 4> staticOffsets(4, 1);
-    SmallVector<int64_t, 4> staticSizes(4, 1);
-    SmallVector<int64_t, 3> staticStrides(3, 1);
-    if (failed(getStaticDims(dmaOp, dmaOp.getSourceMixedOffsets(),
-                             dmaOp.getSourceMixedSizes(),
-                             dmaOp.getSourceMixedStrides(), staticOffsets,
-                             staticSizes, staticStrides))) {
-      return failure();
+    SmallVector<int64_t, 4> staticOffsets;
+    SmallVector<int64_t, 4> staticSizes;
+    SmallVector<int64_t, 3> staticStrides;
+    if (dmaOp.hasSourceAddressing()) {
```

Comment:
Is this underscore prefix a common pattern for variable naming in MLIR/LLVM codebases? I would avoid it as there are some C++ caveats around naming collisions with prefix underscores.

Reply:
Also, why are these additional variables with leading underscores needed? Can't we just work with `staticOffsets` etc directly?

Reply:
Yes, good point - I shall address that.